README.md
# tailwind-extensions

[![npm](https://img.shields.io/npm/v/tailwind-extensions)](https://www.npmjs.com/package/tailwind-extensions)

[![GitHub license](https://img.shields.io/github/license/ModyQyW/tailwind-extensions)](https://github.com/ModyQyW/tailwind-extensions/blob/main/LICENSE)

English | [简体中文](./README.zh-hans.md)

Opinionated TailwindCSS v3 extensions. Requires `node >= 14.18`.

- TailwindCSS v3 enhanced preset for default config.
- TailwindCSS v3 and [Ant Design](https://ant.design/) compat preset
- TailwindCSS v3 and [element-plus](https://element-plus.org/) compat preset
- TailwindCSS v3 and miniprogram compat preset
- TailwindCSS v3 and easy mode compat preset

## Usage

```sh
npm install tailwind-extensions
```

Then update your TailwindCSS config.

### Presets

#### `basePreset`

This preset enhances TailwindCSS default config of TailwindCSS. With this preset you can reduce the use of arbitrary values to keep your attention focused and improve code cleanliness.

```js
const { basePreset } = require('tailwind-extensions');

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [], // set your content
  presets: [basePreset],
};
```

Check out [UnoCss](https://github.com/unocss/unocss) if you have a stronger need for arbitrary values.

#### `antDesignPreset`

This preset provides `Ant Design` v5 colors compat config.

```js
const { basePreset, antDesignPreset } = require('tailwind-extensions');

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [], // set your content
  presets: [
    basePreset,
    antDesignPreset({
      // Adjust the value you need
      /** base */
      baseSelectors: [':root'],
      baseMediaQuery: '',

      basePrimary: '#1677ff',
      baseSecondary: '#666666',
      baseSuccess: '#52c41a',
      baseWarning: '#faad14',
      baseError: '#ff4d4f',
      baseDanger: '#ff4d4f',
      baseInfo: '#1677ff',

      baseBg: '#ffffff',
      baseMaskBg: 'rgba(0, 0, 0, 0.45)',
      baseDisabledBg: 'rgba(0, 0, 0, 0.04)',

      baseBorder: '#d9d9d9',

      baseText: 'rgba(0, 0, 0, 0.85)',
      basePrimaryText: 'rgba(0, 0, 0, 0.85)',
      baseSecondaryText: 'rgba(0, 0, 0, 0.45)',
      baseDisabledText: 'rgba(0, 0, 0, 0.25)',

      baseBoxShadow: '0 1px 2px 0 rgba(0, 0, 0, 0.03), 0 1px 6px -1px rgba(0, 0, 0, 0.02), 0 2px 4px 0 rgba(0, 0, 0, 0.02)',

      /** dark */
      darkSelectors: ['.dark'],
      darkMediaQuery: '',

      darkPrimary: '#1668dc',
      darkSecondary: '#5a5a5a',
      darkSuccess: '#49aa19',
      darkWarning: '#d89614',
      darkError: '#dc4446',
      darkDanger: '#dc4446',
      darkInfo: '#1668dc',

      darkBg: '#141414',
      darkMaskBg: 'rgba(0, 0, 0, 0.45)',
      darkDisabledBg: 'rgba(255, 255, 255, 0.08)',

      darkBorder: '#434343',

      darkText: 'rgba(255, 255, 255, 0.85)',
      darkPrimaryText: 'rgba(255, 255, 255, 0.85)',
      darkSecondaryText: 'rgba(255, 255, 255, 0.45)',
      darkDisabledText: 'rgba(255, 255, 255, 0.3)',

      darkBoxShadow: '0 1px 2px 0 rgba(0, 0, 0, 0.03), 0 1px 6px -1px rgba(0, 0, 0, 0.02), 0 2px 4px 0 rgba(0, 0, 0, 0.02)',
    }),
  ],
};
```

When using this preset, you need to control when to add `class="dark"` to the page container element. [usehooks-ts useDarkMode](https://usehooks-ts.com/react-hook/use-dark-mode) and [VueUse useDark](https://vueuse.org/core/usedark/) are recommended.

`@tailwind base` styles may conflict with the styles associated with the `Ant Design` specification. Please refer to the example below for adjustments.

```ts
// project entry main.ts

// custom preflight 1
import './styles/preflight1.css';

// third-party preflight
import 'modern-normalize';

// TailwindCSS base
import './styles/tailwind-base.css';

// custom preflight 2
import './styles/preflight2.css';

// antd preflight styles
import 'antd/dist/reset.css';

// TailwindCSS components + utilities
// maybe override antd / ant-design-vue preflight if move TailwindCSS base here
import './styles/tailwind.css';

// any other global styles you need
import './styles/global.css';
```

```css
/* styles/tailwind-base.css */
@tailwind base;
```

```css
/* styles/preflight2.css */
html,
page {
  font-size: var(--font-size, 16px);
  line-height: 1.5;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
```

```css
/* styles/tailwind.css */
@tailwind components;
@tailwind utilities;
```

#### `antDesignV4Preset`

This preset provides `Ant Design` v4 colors compat config.

```js
const { basePreset, antDesignV4Preset } = require('tailwind-extensions');

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [], // set your content
  presets: [
    basePreset,
    antDesignV4Preset({
      // Adjust the value you need
      /** base */
      baseSelectors: [':root'],
      baseMediaQuery: '',

      basePrimary: '#1890ff',
      baseSecondary: '#666666',
      baseSuccess: '#52c41a',
      baseWarning: '#faad14',
      baseError: '#f5222d',
      baseDanger: '#f5222d',
      baseInfo: '#1890ff',

      baseBg: '#ffffff',
      baseMaskBg: 'rgba(0, 0, 0, 0.45)',
      baseDisabledBg: '#f5f5f5',

      baseBorder: '#d9d9d9',

      baseText: 'rgba(0, 0, 0, 0.85)',
      basePrimaryText: 'rgba(0, 0, 0, 0.85)',
      baseSecondaryText: 'rgba(0, 0, 0, 0.45)',
      baseDisabledText: 'rgba(0, 0, 0, 0.25)',

      baseBoxShadow: '0 3px 6px -4px rgba(0, 0, 0, 0.12), 0 6px 16px 0 rgba(0, 0, 0, 0.08), 0 9px 28px 8px rgba(0, 0, 0, 0.05)',

      /** dark */
      darkSelectors: ['.dark'],
      darkMediaQuery: '',

      darkPrimary: '#177ddc',
      darkSecondary: '#5a5a5a',
      darkSuccess: '#49aa19',
      darkWarning: '#d89614',
      darkError: '#d32029',
      darkDanger: '#d32029',
      darkInfo: '#177ddc',

      darkBg: '#141414',
      darkMaskBg: 'rgba(0, 0, 0, 0.45)',
      darkDisabledBg: 'rgba(255, 255, 255, 0.08)',

      darkBorder: '#434343',

      darkText: 'rgba(255, 255, 255, 0.85)',
      darkPrimaryText: 'rgba(255, 255, 255, 0.85)',
      darkSecondaryText: 'rgba(255, 255, 255, 0.45)',
      darkDisabledText: 'rgba(255, 255, 255, 0.3)',

      darkBoxShadow: '0 3px 6px -4px rgba(0, 0, 0, 0.48), 0 6px 16px 0 rgba(0, 0, 0, 0.32), 0 9px 28px 8px rgba(0, 0, 0, 0.2)',
    }),
  ],
};
```

When using this preset, you need to control when to add `class="dark"` to the page container element. [usehooks-ts useDarkMode](https://usehooks-ts.com/react-hook/use-dark-mode) and [VueUse useDark](https://vueuse.org/core/usedark/) are recommended.

`@tailwind base` styles may conflict with the styles associated with the `Ant Design` specification. Please refer to the example below for adjustments.

```ts
// project entry main.ts

// custom preflight 1
import './styles/preflight1.css';

// third-party preflight
import 'modern-normalize';

// TailwindCSS base
import './styles/tailwind-base.css';

// custom preflight 2
import './styles/preflight2.css';

// antd styles
import 'antd/dist/antd.min.css';
// import 'antd/dist/antd.variable.min.css';

// ant-design-vue styles
import 'ant-design-vue/dist/antd.min.css';
// import 'ant-design-vue/dist/antd.variable.min.css';

// TailwindCSS components + utilities
// maybe override antd / ant-design-vue preflight if move TailwindCSS base here
import './styles/tailwind.css';

// any other global styles you need
import './styles/global.css';
```

```css
/* styles/tailwind-base.css */
@tailwind base;
```

```css
/* styles/preflight2.css */
html,
page {
  font-size: var(--font-size, 16px);
  line-height: 1.5;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
```

```css
/* styles/tailwind.css */
@tailwind components;
@tailwind utilities;
```

#### `elementPlusPreset`

This preset provides `element-plus` colors compat config.

```js
const { basePreset, elementPlusPreset} = require('tailwind-extensions');

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [], // set your content
  presets: [
    basePreset,
    elementPlusPreset({
      // Adjust the value you need
      /** base */
      baseSelectors: [':root'],
      baseMediaQuery: '',

      basePrimary: '#409eff',
      baseSecondary: '#909399',
      baseSuccess: '#67c23a',
      baseWarning: '#e6a23c',
      baseError: '#f56c6c',
      baseDanger: '#f56c6c',
      baseInfo: '#909399',

      baseFill: '#f0f2f5',
      baseDarkerFill: '#e6e8eb',
      baseDarkFill: '#ebedf0',
      baseLightFill: '#f5f7fa',
      baseLighterFill: '#fafafa',
      baseExtraLightFill: '#fafcff',
      baseBlankFill: '#ffffff',

      baseText: '#303133',
      basePrimaryText: '#303133',
      baseRegularText: '#606266',
      baseSecondaryText: '#909399',
      basePlaceholderText: '#a8abb2',
      baseDisabledText: '#c0c4cc',

      baseBg: '#ffffff',
      basePageBg: '#f2f3f5',
      baseOverlayBg: '#ffffff',
      baseMaskBg: 'rgba(0, 0, 0, 0.5)',
      baseDisabledBg: '#f5f7fa',

      baseBorder: '#dcdfe6',
      baseDarkerBorder: '#cdd0d6',
      baseDarkBorder: '#d4d7de',
      baseLightBorder: '#e4e7ed',
      baseLighterBorder: '#ebeef5',
      baseExtraLightBorder: '#f2f6fc',

      baseBoxShadow: '0px 12px 32px 4px rgba(0, 0, 0, 0.04), 0px 8px 20px rgba(0, 0, 0, 0.08)',

      /** dark */
      darkSelectors: ['.dark'],
      darkMediaQuery: '',

      darkPrimary: '#409eff',
      darkSecondary: '#909399',
      darkSuccess: '#67c23a',
      darkWarning: '#e6a23c',
      darkError: '#f56c6c',
      darkDanger: '#f56c6c',
      darkInfo: '#909399',

      darkFill: '#303030',
      darkDarkerFill: '#424243',
      darkDarkFill: '#39393a',
      darkLightFill: '#262727',
      darkLighterFill: '#1d1d1d',
      darkExtraLightFill: '#191919',
      darkBlankFill: 'transparent',

      darkText: '#e5eaf3',
      darkPrimaryText: '#e5eaf3',
      darkRegularText: '#cfd3dc',
      darkSecondaryText: '#a3a6ad',
      darkPlaceholderText: '#8d9095',
      darkDisabledText: '#6c6e72',

      darkBg: '#141414',
      darkPageBg: '#0a0a0a',
      darkOverlayBg: '#1d1e1f',
      darkMaskBg: 'rgba(0, 0, 0, 0.5)',
      darkDisabledBg: 'transparent',

      darkBorder: '#4c4d4f',
      darkDarkerBorder: '#636466',
      darkDarkBorder: '#58585b',
      darkLightBorder: '#414243',
      darkLighterBorder: '#363637',
      darkExtraLightBorder: '#2b2b2c',

      darkBoxShadow: '0px 12px 32px 4px rgba(0, 0, 0, 0.36), 0px 8px 20px rgba(0, 0, 0, 0.72)',
    }),
  ],
};
```

When using this preset, you need to control when to add `class="dark"` to the page container element. [VueUse useDark](https://vueuse.org/core/usedark/) is recommended.

`@tailwind base` styles may conflict with the `element-plus` styles. Please refer to the example below for adjustments.

```ts
// project entry main.ts

// custom preflight 1
import './styles/preflight1.css';

// third-party preflight
import 'modern-normalize';

// TailwindCSS base
import './styles/tailwind-base.css';

// custom preflight 2
import './styles/preflight2.css';

// element-plus styles
import 'element-plus/dist/index.css';

// TailwindCSS components + utilities
// maybe override element-plus preflight if move TailwindCSS base here
import './styles/tailwind.css';

// any other global styles you need
import './styles/global.css';
```

```css
/* styles/tailwind-base.css */
@tailwind base;
```

```css
/* styles/preflight2.css */
html,
page {
  font-size: var(--font-size, 16px);
  line-height: 1.5;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
```

```css
/* styles/tailwind.css */
@tailwind components;
@tailwind utilities;
```

#### `miniprogramBasePreset`

This preset enhances TailwindCSS default config of TailwindCSS, providing key-value pairs related to `rpx`, only for miniprogram.

```js
const { basePreset, miniprogramBasePreset } = require('tailwind-extensions');

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [], // set your content
  presets: [
    basePreset,
    miniprogramBasePreset,
  ],
};
```

Be sure to include `page` in `baseSelectors` if you use another preset and that preset can accept `baseSelectors` for configuration, as miniprogram may not recognize the `:root` style.

`@tailwind base` styles cannot run in miniprogram. Please use [mini-program-tailwind](https://github.com/dcasia/mini-program-tailwind), [@uni-helper/vite-plugin-uni-tailwind](https://github.com/uni-helper/vite-plugin-uni-tailwind) or others to deal with it.

#### `easyPreset`

This preset provides easy mode compat config.

```js
const { basePreset, easyPreset } = require('tailwind-extensions');

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [], // set your content
  presets: [
    basePreset,
    easyPreset({
      selectors: ['.easy'],
      mediaQuery: '',
      fontSize: '24px',
    }),
  ],
};
```

When using this preset, you need to control when to add `class="easy"` to the page container element. Refer to [usehooks-ts useDarkMode](https://usehooks-ts.com/react-hook/use-dark-mode) and [VueUse useDark](https://vueuse.org/core/usedark/).

To make sure the preset takes effect, please refer to the example below for adjustments.

```ts
// project entry main.ts

// custom preflight 1
import './styles/preflight1.css';

// third-party preflight
import 'modern-normalize';

// TailwindCSS base
import './styles/tailwind-base.css';

// custom preflight 2
import './styles/preflight2.css';

// UI library styles
// import 'xx/yy.css';

// TailwindCSS base + components + utilities
// maybe override UI library preflight if move TailwindCSS base here
import './styles/tailwind.css';

// any other global styles you need
import './styles/global.css';
```

```css
/* styles/tailwind-base.css */
@tailwind base;
```

```css
/* styles/preflight2.css */
html,
page {
  /* Should add !important in miniprogram to make sure it works */
  font-size: var(--font-size, 16px);
  line-height: 1.5;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
```

```css
/* styles/tailwind.css */
@tailwind components;
@tailwind utilities;
```

## [Sponsors](https://github.com/ModyQyW/sponsors)

<p align="center">
  <a href="https://cdn.jsdelivr.net/gh/ModyQyW/sponsors/sponsorkit/sponsors.svg">
    <img src="https://cdn.jsdelivr.net/gh/ModyQyW/sponsors/sponsorkit/sponsors.svg"/>
  </a>
</p>


CHANGELOG.md
# Changelog

## 0.22.1 (2023-03-21)

- fix: fix `element-plus` backgroundColor key

## 0.22.0 (2023-03-08)

- feat: extends more

## 0.21.2 (2023-02-07)

- build: switch to `unbuild`
- fix: fix target

## 0.21.1 (2023-01-04)

- fix: fix exports

## 0.21.0 (2023-01-04)

- build: switch to `rollup`
- feat!: require `tailwindcss@^3.2.0`

## 0.20.0 (2022-12-21)

- feat!: require `node>=14.18`
- feat!: `antDesignPreset` -> `antDesignV4Preset`
- feat: add `antDesignPreset`

## 0.19.2 (2022-11-02)

- fix: fix types

## 0.19.1 (2022-11-02)

- fix: fix `element-plus` colors

## 0.19.0 (2022-11-02)

- feat!: update `aspectRatio` config, following [here](https://github.com/tailwindlabs/tailwindcss-aspect-ratio#compatibility-with-default-aspect-ratio-utilities)

## 0.18.0 (2022-10-14)

- feat!: `tailwind-extensions` -> `tailwind-extensions`
- feat!: update exports
  - `base` -> `basePreset`
  - `antDesign` -> `antDesignPreset`
  - `elementPlus` -> `elementPlusPreset`
  - `miniprogramBase` -> `miniprogramBasePreset`
- feat: update `basePreset`
- feat!: update `antDesignPreset`
  - `baseSelectors` default value `[':root', 'page']` -> `[':root']`
- feat!: update `elementPlusPreset`
  - `baseSelectors` default value `[':root', 'page']` -> `[':root']`
- feat!: update `miniprogramPreset`
  - No longer set key-value pairs related to special symbols like `.` and `/`. Use [mini-program-tailwind](https://github.com/dcasia/mini-program-tailwind) and [@uni-helper/vite-plugin-uni-tailwind](https://github.com/uni-helper/vite-plugin-uni-tailwind) if you need.
- feat!: remove `miniprogramScreensPreset`
- feat!: remove `miniprogramSeparatorPreset`
  - Use [mini-program-tailwind](https://github.com/dcasia/mini-program-tailwind) and [@uni-helper/vite-plugin-uni-tailwind](https://github.com/uni-helper/vite-plugin-uni-tailwind) if you need.
- fix: fix build
- fix: fix `easyPreset` empty theme

## 0.17.0

- feat: extend `aspect-ratio` configs

## 0.16.0

- feat: add more classes

## 0.15.2

- fix: fix docs

## 0.15.1

- fix: fix exports

## 0.15.0

- feat: update exports
- feat!: update `miniprogram` related presets

## 0.14.0

- feat!: split `miniprogram` and `miniprogramEnhanced` presets
- feat: change `baseSelectors` default value to `[':root', 'page']`

## 0.13.0

- feat!: update every preset's usage, especially `miniprogram`

## 0.12.1

## 0.12.0

- feat: update `miniprogram-base.css` (based on ress.css v5.0.2, modern-normalize v1.1.0 and tailwindcss v3.1.5)

## 0.11.2

- fix: fix `type: "module"` support

## 0.11.1

- fix: fix types
- fix: target `node12`

## 0.11.0

- feat: update presets' config
- feat!: tailwindcss@^3.1.0 is required
- fix: fix miniprogram colors

## 0.10.0

- feat: update presets' config

## 0.9.1 (2022-05-23)

## 0.9.0 (2022-05-20)

- feat: don't disable `preflight` by default in `ant-design` and `element-plus` presets

## 0.8.6 (2022-05-09)

- fix: set `darkMediaQuery` to empty string
- fix: fix `miniprogram-base.css`

## 0.8.5 (2022-05-09)

- fix: fix presets' colors

## 0.8.4 (2022-05-09)

- fix: fix presets' colors

## 0.8.3 (2022-05-09)

- fix: fix release hooks

## 0.8.2 (2022-05-09)

- fix: set `darkMode` and `safelist` avoiding wrong purge

## 0.8.1 (2022-05-09)

- fix: fix presets' config

## 0.8.0 (2022-05-09)

- feat: update `miniprogram-base.css`
- fix: fix presets' config
- fix: fix exports

## 0.7.0 (2022-05-07)

- fix: set `ant-design` `darkSelectors` default value to `[.dark]`
- feat: support `element-plus` dark mode
- feat: add `easy` preset

## 0.6.0 (2022-05-03)

- chore: use `tsup` to build and migrate to TypeScript
- fix: fix config

## 0.5.0 (2022-05-01)

- fix: fix docs

## 0.4.0 (2022-04-30)

- feat: update `ant-design` and `element-plus` presets, using `tailwindcss-theme-swapper` to support CSS variables for different themes

## 0.3.0 (2022-04-29)

- feat: update `ant-design` and `element-plus` presets' colors

## 0.2.0 (2022-04-28)

- feat: add `base` preset
- feat: update `ant-design` and `element-plus` presets' default border color
- feat: update `miniprogram` preset

## 0.1.0 (2022-04-27)

- feat: add `ant-design` preset
- feat: add `element-plus` preset
- feat: add `miniprogram` preset
- feat: add `miniprogram-base.css`
- feat: add `miniprogram-preflight.css`


ant-design.js
const themeSwapper = require('tailwindcss-theme-swapper');
const { generate, presetPalettes, presetDarkPalettes } = require('@ant-design/colors');

const convert = (colors) => {
  return {
    DEFAULT: colors.primary || colors[5],
    1: colors[0],
    2: colors[1],
    3: colors[2],
    4: colors[3],
    5: colors[4],
    6: colors[5],
    7: colors[6],
    8: colors[7],
    9: colors[8],
    10: colors[9],
    hover: colors[4],
    active: colors[6],
    'lighten-5': colors[0],
    'light-1': colors[0],
    'lighten-4': colors[1],
    'light-2': colors[1],
    'lighten-3': colors[2],
    'light-3': colors[2],
    'lighten-2': colors[3],
    'light-4': colors[3],
    'lighten-1': colors[4],
    'light-5': colors[4],
    'darken-1': colors[6],
    'dark-1': colors[6],
    'darken-2': colors[7],
    'dark-2': colors[7],
    'darken-3': colors[8],
    'dark-3': colors[8],
    'darken-4': colors[9],
    'dark-4': colors[9],
  };
};

/** @type {import('@types/tailwindcss/tailwind-config').TailwindConfig} */
module.exports = ({
  selectors = [':root'],
  mediaQuery = [],
  primary = '#1890ff',
  secondary = '#666666',
  success = '#52c41a',
  warning = '#faad14',
  error = '#f5222d',
  danger = '#f5222d',
  info = '#1890ff',
  text = 'rgba(0, 0, 0, 0.85)',
  primaryText = 'rgba(0, 0, 0, 0.85)',
  secondaryText = 'rgba(0, 0, 0, 0.45)',
  disabledText = 'rgba(0, 0, 0, 0.25)',
  bg = '#ffffff',
  disabledBg = '#f5f5f5',
  border = '#d9d9d9',
  mask = 'rgba(0, 0, 0, 0.45)',
  boxShadow = '0 3px 6px -4px rgba(0, 0, 0, 0.12), 0 6px 16px 0 rgba(0, 0, 0, 0.08), 0 9px 28px 8px rgba(0, 0, 0, 0.05)',
  darkSelectors = ['.dark', '[dark]', '[data-theme="dark"]', '.is-dark', '[is-dark]'],
  darkMediaQuery = '@media (prefers-color-scheme: dark)',
  darkPrimary = '#177ddc',
  darkSecondary = '#5a5a5a',
  darkSuccess = '#49aa19',
  darkWarning = '#d89614',
  darkError = '#d32029',
  darkDanger = '#d32029',
  darkInfo = '#177ddc',
  darkText = 'rgba(255, 255, 255, 0.85)',
  darkPrimaryText = 'rgba(255, 255, 255, 0.85)',
  darkSecondaryText = 'rgba(255, 255, 255, 0.45)',
  darkDisabledText = 'rgba(255, 255, 255, 0.3)',
  darkBg = '#141414',
  darkDisabledBg = 'rgba(255, 255, 255, 0.08)',
  darkBorder = '#434343',
  darkMask = 'rgba(0, 0, 0, 0.45)',
  darkBoxShadow = '0 3px 6px -4px rgba(0, 0, 0, 0.48), 0 6px 16px 0 rgba(0, 0, 0, 0.32), 0 9px 28px 8px rgba(0, 0, 0, 0.2)',
} = {}) => ({
  corePlugins: {
    preflight: false,
  },
  plugins: [
    themeSwapper({
      themes: [
        {
          name: 'base',
          selectors,
          mediaQuery,
          theme: {
            colors: {
              primary: convert(generate(primary, { theme: 'default', backgroundColor: bg })),

              secondary: convert(generate(secondary, { theme: 'default', backgroundColor: bg })),

              success: convert(generate(success, { theme: 'default', backgroundColor: bg })),

              warning: convert(generate(warning, { theme: 'default', backgroundColor: bg })),

              error: convert(generate(error, { theme: 'default', backgroundColor: bg })),

              danger: convert(generate(danger, { theme: 'default', backgroundColor: bg })),

              info: convert(generate(info, { theme: 'default', backgroundColor: bg })),

              red: convert(presetPalettes.red),

              volcano: convert(presetPalettes.volcano),

              orange: convert(presetPalettes.orange),

              gold: convert(presetPalettes.gold),

              yellow: convert(presetPalettes.yellow),

              lime: convert(presetPalettes.lime),

              green: convert(presetPalettes.green),

              cyan: convert(presetPalettes.cyan),

              blue: convert(presetPalettes.blue),

              geekblue: convert(presetPalettes.geekblue),

              purple: convert(presetPalettes.purple),

              magenta: convert(presetPalettes.magenta),

              gray: {
                DEFAULT: '#8c8c8c',
                1: '#ffffff',
                2: '#fafafa',
                3: '#f5f5f5',
                4: '#f0f0f0',
                5: '#d9d9d9',
                6: '#bfbfbf',
                7: '#8c8c8c',
                8: '#595959',
                9: '#434343',
                10: '#262626',
                11: '#1f1f1f',
                12: '#141414',
                13: '#000000',
                hover: '#bfbfbf',
                active: '#595959',
                'lighten-6': '#ffffff',
                'light-6': '#ffffff',
                'lighten-5': '#fafafa',
                'light-5': '#fafafa',
                'lighten-4': '#f5f5f5',
                'light-4': '#f5f5f5',
                'lighten-3': '#f0f0f0',
                'light-3': '#f0f0f0',
                'lighten-2': '#d9d9d9',
                'light-2': '#d9d9d9',
                'lighten-1': '#bfbfbf',
                'light-1': '#bfbfbf',
                'darken-1': '#595959',
                'dark-1': '#595959',
                'darken-2': '#434343',
                'dark-2': '#434343',
                'darken-3': '#262626',
                'dark-3': '#262626',
                'darken-4': '#1f1f1f',
                'dark-4': '#1f1f1f',
                'darken-5': '#141414',
                'dark-5': '#141414',
                'darken-6': '#000000',
                'dark-6': '#000000',
              },

              grey: {
                DEFAULT: '#8c8c8c',
                1: '#ffffff',
                2: '#fafafa',
                3: '#f5f5f5',
                4: '#f0f0f0',
                5: '#d9d9d9',
                6: '#bfbfbf',
                7: '#8c8c8c',
                8: '#595959',
                9: '#434343',
                10: '#262626',
                11: '#1f1f1f',
                12: '#141414',
                13: '#000000',
                hover: '#bfbfbf',
                active: '#595959',
                'lighten-6': '#ffffff',
                'light-6': '#ffffff',
                'lighten-5': '#fafafa',
                'light-5': '#fafafa',
                'lighten-4': '#f5f5f5',
                'light-4': '#f5f5f5',
                'lighten-3': '#f0f0f0',
                'light-3': '#f0f0f0',
                'lighten-2': '#d9d9d9',
                'light-2': '#d9d9d9',
                'lighten-1': '#bfbfbf',
                'light-1': '#bfbfbf',
                'darken-1': '#595959',
                'dark-1': '#595959',
                'darken-2': '#434343',
                'dark-2': '#434343',
                'darken-3': '#262626',
                'dark-3': '#262626',
                'darken-4': '#1f1f1f',
                'dark-4': '#1f1f1f',
                'darken-5': '#141414',
                'dark-5': '#141414',
                'darken-6': '#000000',
                'dark-6': '#000000',
              },

              text: {
                DEFAULT: text,
                primary: primaryText,
                secondary: secondaryText,
                disabled: disabledText,
              },

              bg: {
                DEFAULT: bg,
                disabled: disabledBg,
              },

              border,

              mask,
            },

            boxShadow,
          },
        },
        {
          name: 'dark',
          selectors: darkSelectors,
          mediaQuery: darkMediaQuery,
          theme: {
            colors: {
              primary: convert(generate(darkPrimary, { theme: 'dark', backgroundColor: darkBg })),

              secondary: convert(
                generate(darkSecondary, { theme: 'dark', backgroundColor: darkBg }),
              ),

              success: convert(generate(darkSuccess, { theme: 'dark', backgroundColor: darkBg })),

              warning: convert(generate(darkWarning, { theme: 'dark', backgroundColor: darkBg })),

              error: convert(generate(darkError, { theme: 'dark', backgroundColor: darkBg })),

              danger: convert(generate(darkDanger, { theme: 'dark', backgroundColor: darkBg })),

              info: convert(generate(darkInfo, { theme: 'dark', backgroundColor: darkBg })),

              red: convert(presetDarkPalettes.red),

              volcano: convert(presetDarkPalettes.volcano),

              orange: convert(presetDarkPalettes.orange),

              gold: convert(presetDarkPalettes.gold),

              yellow: convert(presetDarkPalettes.yellow),

              lime: convert(presetDarkPalettes.lime),

              green: convert(presetDarkPalettes.green),

              cyan: convert(presetDarkPalettes.cyan),

              blue: convert(presetDarkPalettes.blue),

              geekblue: convert(presetDarkPalettes.geekblue),

              purple: convert(presetDarkPalettes.purple),

              magenta: convert(presetDarkPalettes.magenta),

              text: {
                DEFAULT: darkText,
                primary: darkPrimaryText,
                secondary: darkSecondaryText,
                disabled: darkDisabledText,
              },

              bg: {
                DEFAULT: darkBg,
                disabled: darkDisabledBg,
              },

              border: darkBorder,

              mask: darkMask,
            },

            boxShadow: darkBoxShadow,
          },
        },
      ],
    }),
  ],
  theme: {
    screens: {
      xs: '480px',
      sm: '576px',
      md: '768px',
      lg: '992px',
      xl: '1200px',
      xxl: '1600px',
      '2xl': '1600px',
    },
    extend: {
      borderColor: {
        DEFAULT: 'var(--colors-border, currentColor)',
      },
      fontFamily: {
        font: "-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, 'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji'",
      },
    },
  },
});
